docs/website/docs/general-usage/schema-evolution.md

```diff
@@ -0,0 +1,205 @@
+
+**Getting notifications**
+We can read the load outcome and send it to slack webhook with dlt.
+```python
+# Import the send_slack_message function from the dlt library
+from dlt.common.runtime.slack import send_slack_message
+
+# Define the URL for your Slack webhook
+hook = "https://hooks.slack.com/services/xxx/xxx/xxx"
+
+# Iterate over each package in the load_info object
+for package in info.load_packages:
+    # Iterate over each table in the schema_update of the current package
+    for table_name, table in package.schema_update.items():
+        # Iterate over each column in the current table
+        for column_name, column in table["columns"].items():
+            # Send a message to the Slack channel with the table 
+						# and column update information
+            send_slack_message(
+                hook,
+                message=(
+                    f"\tTable updated: {table_name}: "
+                    f"Column changed: {column_name}: "
+                    f"{column['data_type']}"
+                )
+            )
+```
+This script sends Slack notifications for data schema updates using the 'send_slack_message' function from the dlt library. It provides details on the updated table and column.
+
+## How to control evolution
+
+`dlt` allows schema evolution control via its schema and data contracts. Refer to our **[documentation](https://dlthub.com/docs/general-usage/schema-contracts)** for details.
+
+### How to test for removed columns - applying “not null” constraint
+
+A column not existing, and a column being null, are two different things. However, when it comes to APIs and json, it’s usually all treated the same - the key-value pair will simply not exist. 
+
+To remove a column, exclude it from the output of the resource function. Subsequent data inserts will treat this column as null. Verify column removal by applying a not null constraint. For instance, after removing the "room" column, apply a not null constraint to confirm its exclusion.
+
+```python
+import dlt
+
+# Define a data resource using dlt.resource with a schema contract
+@dlt.resource(schema_contract={"tables": "evolve", "columns": "evolve"}, columns={"room": {"data_type": "integer", "nullable": "false"}})
+def data_resource():
+    yield {
+    "organization": "Tech Innovations Inc.",
+    "address": {
+        'building': 'r&d' 
+        #"room": 7890,
+    },
+    "Inventory": [
+        {"name": "Plasma ray", "inventory nr": 2411},
+        {"name": "Self-aware Roomba", "inventory nr": 268},
+        {"name": "Type-inferrer", "inventory nr": 3621}
+    ]
+}
+```
+During pipeline execution a data validation error indicates that a removed column is being passed as null.
+
+## Some schema changes in the data
+
+The data in the pipeline mentioned above is modified.
+
+- Updated data pipeine now includes key 'specifications' within 'details', nested in 'Inventory'.
+```python
```

Comment:
delete please extra spaces in the code snippet

Reply:
done